File: tui/packages/ui/src/components/run-target-info.test.ts
import { describe, expect, test } from 'bun:test';
import type { App, TableRow } from '@devenv/types';
import { appRunTargetDetailRows } from './AppDetailView';
import { appRunTargetRightMetadata } from './Table';

const baseApp: App = {
  ident: 'app',
  displayName: 'App',
  localDirectoryPath: '/repo/app',
  repositoryPath: '/repo/app',
  branch: 'main',
  appType: 'APP',
  containerBaseName: 'app',
};

describe('run target info rendering helpers', () => {
  test('detail rows include rough running since time', () => {
    const rows = appRunTargetDetailRows({
      ...baseApp,
      runTargetInfo: {
        runtime: 'shell',
        launchMode: 'tmux',
        label: 'bun build',
        profile: 'default',
        targetId: 'app/app/run/shell/default',
        sourcePath: '/repo/app/package.json',
        startedAt: '2026-01-02T03:04:05Z',
        display: '[tmux] bun build (default)',
      },
    });

    expect(rows).toHaveLength(1);
    expect(rows[0].label).toBe('Since');
    expect(rows[0].value).toContain('1/2/2026');
  });

  test('detail rows hidden when missing', () => {
    expect(appRunTargetDetailRows(baseApp)).toEqual([]);
  });

  test('table right metadata helper returns compact hint for apps only', () => {
    const row: TableRow = { ...baseApp, rowKind: 'app', runTargetInfo: { runtime: 'docker', startedAt: '2026-01-02T03:04:05Z', display: '[docker] default (default)' } };
    expect(appRunTargetRightMetadata(row)).toBe('[docker] default (default)');
    expect(appRunTargetRightMetadata({ ...row, runTargetInfo: undefined })).toBeUndefined();
  });
});

File: server/pkg/server/run_target_info_test.go
package server

import (
	"testing"

	"github.com/friendsfriend/devenv/pkg/app"
	"github.com/friendsfriend/devenv/pkg/build"
	"github.com/friendsfriend/devenv/pkg/docker"
	"github.com/friendsfriend/devenv/pkg/resources"
	"github.com/friendsfriend/devenv/pkg/runstatus"
)

func TestAppStatusEventPropertiesRunTargetInfoOptional(t *testing.T) {
	buildSvc := build.NewService(nil, nil, nil, "")
	buildSvc.SetRunTargetInfo("app", resources.ActionTarget{ID: "target", Runtime: resources.ActionRuntimeShell, LaunchMode: resources.LaunchModeTmux, Label: "bun build", Profile: "default", SourcePath: "/repo/package.json"})
	withInfo := &Server{services: &stubServicesContainer{buildService: buildSvc}}
	runtimeStatus := runstatus.Normalize("running")
	props := withInfo.appStatusEventProperties(app.App{Ident: "app", AppType: app.TypeAPP}, &docker.Info{}, "✓", "main", nil, &runtimeStatus)
	info, ok := props["runTargetInfo"].(*build.RunTargetInfo)
	if !ok || info.Display != "[tmux] bun build (default)" {
		t.Fatalf("runTargetInfo = %#v", props["runTargetInfo"])
	}

	withoutInfo := &Server{services: &stubServicesContainer{buildService: build.NewService(nil, nil, nil, "")}}
	props = withoutInfo.appStatusEventProperties(app.App{Ident: "app", AppType: app.TypeAPP}, &docker.Info{}, "✓", "main", nil, &runtimeStatus)
	if value, ok := props["runTargetInfo"]; !ok || value != nil {
		t.Fatalf("runTargetInfo should clear when absent: %#v", value)
	}
}
